pilot/pkg/model/gateway.go

```diff
@@ -45,24 +56,101 @@ func MergeGateways(gateways ...Config) *MergedGateway {
 		log.Debugf("MergeGateways: merging gateway %q into %v:\n%v", name, names, gateway)
 		for _, s := range gateway.Servers {
 			log.Debugf("MergeGateways: gateway %q processing server %v", name, s.Hosts)
-			if ss, ok := servers[s.Port.Number]; ok {
-				// TODO: remove this check when Envoy supports filter chain matching so we can expose multiple protocols on the same physical port
-				// ss must have at least one element because the key exists in the map, otherwise we'd be in the else case below.
-				if ss[0].Port.Protocol != s.Port.Protocol {
-					log.Debugf("skipping server: attempting to merge servers for gateway %q into %v but servers have different protocols: want %v have %v",
-						spec.Name, names, ss[0].Port.Protocol, s.Port.Protocol)
-					continue
+			protocol := ParseProtocol(s.Port.Protocol)
+			if gatewayPorts[s.Port.Number] {
+				// We have two servers on the same port. Should we merge?
+				// 1. Yes if both servers are plain text and HTTP
+				// 2. Yes if both servers are using TLS
+				//    if using HTTPS ensure that port name is distinct so that we can setup separate RDS
+				//    for each server (as each server ends up as a separate http connection manager due to filter chain match
+				// 3. No for everything else.
+
+				if p, exists := plaintextServers[s.Port.Number]; exists {
+					currentProto := ParseProtocol(p[0].Port.Protocol)
+					if currentProto != protocol || protocol != ProtocolHTTP {
+						log.Debugf("skipping server on gateway %s port %s.%d.%s: conflict with existing server %s.%d.%s",
+							spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol, p[0].Port.Name, p[0].Port.Number, p[0].Port.Protocol)
+						continue
+					}
+					rdsName := getRDSRouteName(s)
+					rdsRouteConfigNames[rdsName] = append(rdsRouteConfigNames[rdsName], s)
+				} else {
+					// This has to be a TLS server
+					if isHTTPServer(s) {
+						rdsName := getRDSRouteName(s)
+						// both servers are HTTPS servers. Make sure the port names are different so that RDS can pick out individual servers
+						// WE cannot have two servers with same port name because we need the port name to distinguish one HTTPS server from another
+						// WE cannot merge two HTTPS servers even if their TLS settings have same path to the keys, because we don't know if the contents
+						// of the keys are same. So we treat them as effectively different TLS settings.
+						if _, exists := rdsRouteConfigNames[rdsName]; exists {
+							log.Debugf("skipping server on gateway %s port %s.%d.%s: non unique port name for HTTPS port",
+								spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol)
+							continue
+						}
+						rdsRouteConfigNames[rdsName] = []*networking.Server{s}
+					}
+					tlsServers[s.Port.Number] = append(tlsServers[s.Port.Number], s)
```

Comment:
Same as above, I don't think we ever explicitly checked that this was a server with TLS, because it could be a plain old HTTP server too, right?